nnvm/python/nnvm/testing/check_computation.py

```diff
@@ -0,0 +1,451 @@
+        list of pairs `(var_name, shape)`.
+
+    np_forward : Callable[..., List[numpy.ndarray]], optional
+        A reference implementation to compare with.
+
+    np_backward : Callable[..., List[numpy.ndarray]], optional
+        A reference implementation of gradients. Should also accept head_grads besides
+        normal inputs. Should return gradients with respect to variables from grad_input_vars in
+        exactly the same order.
+
+    shape : Dict[str, Tuple], optional
+        A dict mapping input variable names to shapes.
+        By default shapes will be inferred automatically.
+
+    dtype : Dict[str, str] or str, optional
+        A dict mapping input variable names to dtypes, or just a single dtype.
+        By default dtypes will be inferred automatically.
+
+    in_range : Dict[str, (float, float)] or (float, float), optional
+        A dict mapping input variable names to ranges or just a single range
+        (the same for all variables). Input values will be generated from
+        uniform distributions on these ranges. `head_grads` can also be
+        assigned a range this way.
+
+    exclude_targets : Set[str]
+        Skip compiling and running anything for these targets.
+
+    only_targets : Set[str]
+        Test only for those targets from `ctx_list()` that are also in this set.
+
+    numerical_grads : bool or "if_possible"
+        Whether to additionally check against numerically computed gradients. If 'if_possible' is
+        passed (which is the default) then it will try to create a gradient computation graph and
+        then check gradients numerically only if this graph can be created (i.e. if there are some
+        operations with unimplemented gradients, it will just issue a warning).
+
+    delta : float
+        A small value used for numerical gradient computation (usually called h in textbooks).
+
+    atol : float
+        Absolute tolerance.
+
+    rtol : float
+        Relative tolerance.
+
+    ng_atol : float
+        Absolute tolerance for numerical grad checking.
+
+    ng_rtol : float
+        Relative tolerance for numerical grad checking.
+
+    ng_max_error : float
+        Discard numerical partial derivatives whose estimated error is larger than this value.
+
+    ng_max_discarded_frac : float
+        Allow discarding no more than this fraction of partial derivatives.
+
+    dump_graph : bool
+        Dump the graph even on success.
+    """
+
+    if numerical_grads not in [False, True, 'if_possible']:
+        raise ValueError("numerical_grads must be a bool or 'if_possible', not {}"
+                         .format(numerical_grads))
+
+    input_vars = [x for x in symbol.list_input_variables()]
+    input_dict = {x.attr('name'): x for x in input_vars}
+
+    if grad_input_vars is None:
+        grad_input_vars = input_vars
+
+    shape = shape.copy() if shape else {}
+
+    grad_input_vars_real = []
+    for x in grad_input_vars:
```

Comment:
This logic looks complicated. How about just passing a list of symbol as grad_input_vars? Shape information can be defined in shape(maybe input_shapes of dict from str to tuple?) argument so that we don't have multiple shapes.